File: BackEnd/src/data/IMappers/IPostMapper.ts
/**
 * @file IPostMapper.ts
 * @description Interfaz del mapper de posts de la comunidad.
 * Define el contrato de transformación entre documento MongoDB, entidad de dominio y DTO de respuesta.
 * @module Community
 * @layer Data
 */

import type { Post } from '../../domain/entities/Post.js';
import type { PostDocument } from '../datasources/mongodb/models/PostModel.js';
import type { PostResponseDTO } from '../../domain/dtos/community/post-response.dto.js';

/**
 * Contrato del mapper de posts.
 * Los repositorios dependen de esta interfaz, no de la implementación concreta.
 */
export interface IPostMapper {
  /**
   * Convierte un documento MongoDB a entidad Post de dominio.
   *
   * @param doc — Documento de la colección 'posts'.
   * @returns Entidad Post.
   */
  toEntity(doc: PostDocument): Post;

  /**
   * Convierte una entidad Post a documento MongoDB (sin _id).
   *
   * @param entity — Entidad Post.
   * @returns Documento parcial para insertar en MongoDB.
   */
  toDocument(entity: Omit<Post, 'id' | 'isActive'>): Omit<PostDocument, '_id'>;

  /**
   * Convierte una entidad Post al DTO de respuesta HTTP, enriquecido con datos del autor.
   *
   * @param entity      — Entidad Post.
   * @param authorName  — Nombre del autor del post.
   * @param authorPhoto — URL de la foto del autor (opcional).
   * @param isLikedByMe — true si el usuario que solicita ya dio like (por defecto false).
   * @returns PostResponseDTO serializable.
   */
  toResponseDTO(entity: Post, authorName: string, authorPhoto?: string, isLikedByMe?: boolean): PostResponseDTO;
}

File: FrontEnd/lib/data/i_mappers/i_post_mapper.dart
/// @file i_post_mapper.dart
/// @description Interfaz del mapper de posts de comunidad. Contrato PostModel ↔ Post.
/// @module Community
/// @layer Data
library;

import '../../domain/entities/post.dart';
import '../models/post_model.dart';

/// Contrato para la conversión bidireccional entre PostModel y Post.
///
/// Los repositorios dependen de esta interfaz para no acoplarse a la
/// implementación concreta del mapper.
abstract interface class IPostMapper {
  /// Convierte un modelo de serialización a entidad de dominio.
  ///
  /// [param] model — PostModel obtenido desde la API o caché.
  /// [returns] Post — Entidad inmutable de dominio.
  Post toEntity(PostModel model);

  /// Convierte una entidad de dominio a modelo de serialización.
  ///
  /// [param] entity — Post de dominio.
  /// [returns] PostModel — Modelo listo para serializar a JSON.
  PostModel toModel(Post entity);
}
